sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -270,19 +306,24 @@ public void processElement(ProcessContext c, BoundedWindow window) throws Except
       KV<BoundedWindow, PaneInfo> key = KV.of(window, paneInfo);
       writer = windowedWriters.get(key);
       if (writer == null) {
-        String uuid = UUID.randomUUID().toString();
-        LOG.info(
-            "Opening writer {} for write operation {}, window {} pane {}",
-            uuid,
-            writeOperation,
-            window,
-            paneInfo);
-        writer = writeOperation.createWriter();
-        writer.openWindowed(uuid, window, paneInfo, UNKNOWN_SHARDNUM);
-        windowedWriters.put(key, writer);
-        LOG.debug("Done opening writer");
+        if (windowedWriters.size() <= maxNumWritersPerBundle) {
+          String uuid = UUID.randomUUID().toString();
+          LOG.info(
+              "Opening writer {} for write operation {}, window {} pane {}",
+              uuid,
+              writeOperation,
+              window,
+              paneInfo);
+          writer = writeOperation.createWriter();
+          writer.openWindowed(uuid, window, paneInfo, UNKNOWN_SHARDNUM);
+          windowedWriters.put(key, writer);
+          LOG.debug("Done opening writer");
+        } else {
+          c.output(unwrittedRecordsTag, KV.of(
```

Comment:
Think making it a separate DoFn is overkill, but the suggestion of caching the random number is a good one (and what we do elsewhere)